File: 39-combination-sum/combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res=new ArrayList<>();
        List<Integer> ll=new ArrayList<>();
        comb(candidates,target,0,res,ll);
        return res;
    }
    public void comb(int[] arr,int target,int idx,List<List<Integer>> res,List<Integer> ll){
        if(target==0){
            res.add(new ArrayList<>(ll));
        }
        for(int i=idx;i<arr.length;i++){
            if(target>=arr[i]){
                ll.add(arr[i]);
                comb(arr,target-arr[i],i,res,ll);
                ll.remove(ll.get(ll.size()-1));
            }
        }
    }
}

File: 39-combination-sum/combination-sum.cpp
class Solution {
public:
    vector<int> ll;
    vector<vector<int>> ans;
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        solve(candidates,target,0);
        return ans;
    }
    void solve(vector<int>& arr,int val,int idx){
        if(val==0){
            ans.push_back(ll);
            return;
        }
        for(int i=idx;i<arr.size();i++){
            if(val>=arr[i]){
                ll.push_back(arr[i]);
                solve(arr,val-arr[i],i);
                ll.pop_back();
            }
        }
    }
};